Use metadata api in *_network_options

spacedmonkey opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket:  https://core.trac.wordpress.org/ticket/37181

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

spacedmonkey committed:
Refresh patch.

spacedmonkey committed:
More performance tweaks.

spacedmonkey committed:
Add some more unit tests.

spacedmonkey committed:
Fix types.

spacedmonkey committed:
Fix lints.

spacedmonkey committed:
Fix types again.

spacedmonkey committed:
Merge branch 'trunk' into ix/37181-a

spacedmonkey committed:
Merge branch 'trunk' into ix/37181-a

manuelRod commented on src/wp-includes/option.php:
```diff
@@ -323,36 +323,20 @@ function wp_load_alloptions( $force_cache = false ) {
 function wp_load_core_site_options( $network_id = null ) {
-	global $wpdb;
-
-	if ( ! is_multisite() || wp_using_ext_object_cache() || wp_installing() ) {
```
Why wp_using_ext_object_cache() is removed?

JJJ replied:
I think that it can safely be removed, because this check would now be handled by the meta-data API instead?

spacedmonkey replied:
The meta api, checks to see if meta is in the object cache and if not then it loads it from the database. One of the many reason I want to make this change, so much of this stuff is handle upstream.

spacedmonkey replied:
For reference https://github.com/WordPress/wordpress-develop/blob/04e9728701b3dde0260c7cdebfd726edbe62ac97/src/wp-includes/meta.php#L1151-L1191

JJJ commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -123,86 +123,52 @@ public function data_network_id_parameter() {
+	 * @ticket 37181
 	 * @group ms-required
 	 */
-	public function test_get_network_option_sets_notoptions_if_option_not_found() {
-		$network_id     = get_current_network_id();
-		$notoptions_key = "$network_id:notoptions";
-
-		$original_cache = wp_cache_get( $notoptions_key, 'site-options' );
-		if ( false !== $original_cache ) {
-			wp_cache_delete( $notoptions_key, 'site-options' );
-		}
-
-		// Retrieve any non-existing option.
-		get_network_option( $network_id, 'this_does_not_exist' );
-
-		$cache = wp_cache_get( $notoptions_key, 'site-options' );
-		if ( false !== $original_cache ) {
-			wp_cache_set( $notoptions_key, $original_cache, 'site-options' );
-		}
-
-		$this->assertSame( array( 'this_does_not_exist' => true ), $cache );
```
I'm guessing it can safely be removed because the meta version of this implementation already has its own tests?

Is there a sane way (does it _possibly_ make sense) to keep this test around but modify its internals, just for the sake of procedural code coverage?

spacedmonkey replied:
notoptions cache is not needed in this implementation. It is removed, so the test should be removed.

JJJ replied:
I understand. 😅

But... what happens (with this PR applied) when `get_network_option()` is called multiple times using an option key that is known not to exist in the database?

Currently, that miss is cached using the `notoptions` cache key to prevent multiple database misses from happening.

Is that still the case, and should it continue to be tested? (I think, probably?)

spacedmonkey replied:
@JJJ I have added a test to prove that if you try to hit network option that does not exist. It does not result in another SQL query, which is neat.

JJJ commented on tests/phpunit/tests/option/multisite.php:
```diff
@@ -131,26 +131,6 @@ public function test_with_another_site() {
 			// $this->assertFalse( get_option( $key2 ) );                        // Check get_option().
 		}
 
-		/**
-		 * @group multisite
-		 */
-		public function test_site_notoptions() {
-			$network_id     = get_current_network_id();
-			$notoptions_key = "{$network_id}:notoptions";
-
-			$_notoptions = wp_cache_get( 'notoptions', 'site-options' );
-			$this->assertEmpty( $_notoptions );
-			$_notoptions1 = wp_cache_get( $notoptions_key, 'site-options' );
-			$this->assertEmpty( $_notoptions1 );
-
-			get_site_option( 'burrito' );
-
-			$notoptions = wp_cache_get( 'notoptions', 'site-options' );
-			$this->assertEmpty( $notoptions );
-			$notoptions1 = wp_cache_get( $notoptions_key, 'site-options' );
-			$this->assertNotEmpty( $notoptions1 );
-		}
-
```
Same as: https://github.com/WordPress/wordpress-develop/pull/2478/commits/938276780a76b5f55d524e823fb80d0fcdc96edc#r864886013

spacedmonkey replied:
Same here, notoptions is not a thing, so we should not test for it.

spacedmonkey commented on src/wp-includes/class-wp-network-query.php:
```diff
@@ -111,27 +111,30 @@ class WP_Network_Query {
 	 *     @type string       $search               Search term(s) to retrieve matching networks for. Default empty.
 	 *     @type bool         $update_network_cache Whether to prime the cache for found networks. Default true.
+	 *     @type bool         $update_network_meta_cache  Whether to prime the metadata (option) cache for found networks.
+	 *                                                    Default false.
```
```suggestion
	 *                                                    Default true.
```

spacedmonkey commented on src/wp-includes/class-wp-network-query.php:
```diff
@@ -111,27 +111,30 @@ class WP_Network_Query {
+			'path__not_in'              => '',
+			'search'                    => '',
+			'update_network_cache'      => true,
+			'update_network_meta_cache' => false,
```
```suggestion
			'update_network_meta_cache' => true,
```

spacedmonkey committed:
Apply suggestions from code review

pbearne reviewed (approved):
A bit to the review But I can't see any problems

spacedmonkey committed:
Merge branch 'trunk' into ix/37181-a

spacedmonkey committed:
No need to double maybe_serialize

spacedmonkey committed:
Upddate 6.1.0

spacedmonkey committed:
Add a new unit test.

spacedmonkey committed:
Update 6.1.0

peterwilsoncc commented on src/wp-includes/load.php:
```diff
@@ -730,7 +730,7 @@ function wp_start_object_cache() {
 
 	if ( function_exists( 'wp_cache_add_global_groups' ) ) {
-		wp_cache_add_global_groups( array( 'users', 'userlogins', 'usermeta', 'user_meta', 'useremail', 'userslugs', 'site-transient', 'site-options', 'blog-lookup', 'blog-details', 'site-details', 'rss', 'global-posts', 'blog-id-cache', 'networks', 'sites', 'blog_meta' ) );
+		wp_cache_add_global_groups( array( 'users', 'userlogins', 'usermeta', 'user_meta', 'useremail', 'userslugs', 'site-transient', 'site-options', 'blog-lookup', 'blog-details', 'site-details', 'rss', 'global-posts', 'blog-id-cache', 'networks', 'sites', 'blog_meta', 'site_meta' ) );
```
I thought `blog_meta` was used for site meta for consistency with the database table name. Please correct me if I am confused about names.

JJJ replied:
> Please correct me if I am confused about names.

Everyone is confused about multisite names because multisite names confuse everyone 🤣

In this specific instance, `blog_meta` refers to the `wp_blogmeta` database table via the meta API, hence `site_meta` will similarly refer to the `wp_sitemeta` database table.

meta.php gathers the cache-group name via a simple concatenation:
```php
$meta_type . '_meta'
```

Because this PR is dropping the single "network options" approach for the "per network meta" one, there is no cache group to drop, and 1 to add.

peterwilsoncc replied:
Thanks John!

I think I have it: blogs are sites, sites are networks, naming things remains difficult.

peterwilsoncc reviewed (commented):
Back when site/blog/whatever meta was introduced the dev note said:

> There is a trade-off between using site options and site metadata, so it is recommended to think about every piece of extra data associated with a site and how it should be stored.
>
> * Use options when the data concerns only the site itself. Using an option should continue to be the default approach for site data.
> * Use site metadata when the data also needs to be heavily used in the network context, especially when sites need to be queried by it.
-- [source](https://make.wordpress.org/core/2019/01/28/multisite-support-for-site-metadata-in-5-1/)

As @felixarntz wrote that, I'll defer to his thoughts.

The affect of this is that all options become `alloptions` so there is some risk of hitting cache object limits for networks with a lot of options.

spacedmonkey commented:
@peterwilsoncc Will confused what what you are getting out there. One mentions blogmeta. There is network options, which is a little different. That difference is important here. 

I understand not want to put everything in once cache key. But network options are a little more controlled than options and alloptions. Autoloading options on sites is a performance issue. But that is when you have 500+ options on a site. See this [link](https://10up.com/blog/2017/wp-options-table/) for more information. As network options are not used as much as they only for network options, network options do not have the same problems as per site options. I think the many benefits here, overcall, this small pitfall.

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -1659,20 +1575,7 @@ function delete_network_option( $network_id, $option ) {
-				'site_id'  => $network_id,
-			)
-		);
+		$result = delete_metadata( 'site', $network_id, $option, '' );
```
```suggestion
		$result = delete_metadata( 'site', $network_id, wp_slash( $option ) );
```

Default not needed (nit pick)

ref:
https://github.com/WordPress/wordpress-develop/blob/deafd1193c244acfae80bb4507de4d8b532fa9e5/src/wp-includes/meta.php#L398

spacedmonkey replied:
👍 Fixed in https://github.com/WordPress/wordpress-develop/pull/2478/commits/faf732ada8e0f3092ffabb3e7d013b778dfade39

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -1536,48 +1489,11 @@ function add_network_option( $network_id, $option, $value ) {
-			wp_cache_set( $notoptions_key, $notoptions, 'site-options' );
-		}
+		$value  = sanitize_option( $option, $value );
+		$result = add_metadata( 'site', $network_id, $option, $value, true );
```
```suggestion
		$result = add_metadata( 'site', $network_id, wp_slash( $option ), wp_slash( $value ), true );
```

ref: https://github.com/WordPress/wordpress-develop/blob/deafd1193c244acfae80bb4507de4d8b532fa9e5/src/wp-includes/meta.php#L52-L53

spacedmonkey replied:
Great call out. I have added unit tests and this fix. https://github.com/WordPress/wordpress-develop/pull/2478/commits/faf732ada8e0f3092ffabb3e7d013b778dfade39

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -1774,33 +1675,11 @@ function update_network_option( $network_id, $option, $value ) {
-			wp_cache_set( $cache_key, $value, 'site-options' );
-		}
+		$value  = sanitize_option( $option, $value );
+		$result = update_metadata( 'site', $network_id, $option, $value );
```
```suggestion
		$result = update_metadata( 'site', $network_id, wp_slash( $option ), wp_slash( $value ) );
```

ref

https://github.com/WordPress/wordpress-develop/blob/deafd1193c244acfae80bb4507de4d8b532fa9e5/src/wp-includes/meta.php#L203-L207

JJJ replied:
🥴

spacedmonkey replied:
Pushed a fix https://github.com/WordPress/wordpress-develop/pull/2478/commits/faf732ada8e0f3092ffabb3e7d013b778dfade39

peterwilsoncc reviewed (commented):
Some memories stirred about the need for slashing meta data.

Otherwise I think this looks good.

spacedmonkey committed:
Merge branch 'trunk' into ix/37181-a

spacedmonkey committed:
Add more unit tests.

spacedmonkey committed:
Fix unit tests running.

peterwilsoncc commented on tests/phpunit/tests/ajax/Compression.php:
```diff
@@ -154,7 +154,7 @@ public function test_set_yes() {
 
 		// Check the site option is not changed due to lack of nonce.
-		$this->assertSame( 0, get_site_option( 'can_compress_scripts' ) );
+		$this->assertSame( 0, (int) get_site_option( 'can_compress_scripts' ) );
```
This is wild (from `trunk`, not your branch!)
![Screen Shot 2022-07-06 at 1 01 36 pm](https://user-images.githubusercontent.com/519727/177459004-cd07c9ff-3662-4729-b488-a8bf824c9332.png)

peterwilsoncc reviewed (approved):
Looks good to me.

I guess the question is should https://core.trac.wordpress.org/ticket/56119 be resolved prior to this going in. I have no idea.

desrosj commented on src/wp-includes/ms-functions.php:
```diff
@@ -113,7 +113,7 @@ function get_active_blog_for_user( $user_id ) {
  */
 function get_blog_count( $network_id = null ) {
-	return get_network_option( $network_id, 'blog_count' );
+	return (int) get_network_option( $network_id, 'blog_count' );
```
Is there a reason that this type casting needs to be added in combination with the other changes? Not against including it, but just wanted to understand a bit why this is included as it _could_ potentially cause problems with someone's code if they're filtering the value and returning an incorrect type. Probably super edge case though.

spacedmonkey replied:
blog_count and get_space_allowed are stored as strings. But somehow become a int here. This change makes them into strings. So casting to int is needed to ensure unit tests and phpdocs do not fail.

spacedmonkey replied:
See https://github.com/WordPress/wordpress-develop/pull/2478#discussion_r914345925

spacedmonkey replied:
Also see https://github.com/WordPress/wordpress-develop/blob/2d9073413fa4714d13cc92075fcf4b9d7318777d/src/wp-includes/user.php#L1338

desrosj commented on src/wp-includes/ms-functions.php:
```diff
@@ -2654,7 +2654,7 @@ function get_space_allowed() {
 	 * @param int $space_allowed Upload quota in megabytes for the current blog.
 	 */
-	return apply_filters( 'get_space_allowed', $space_allowed );
+	return (int) apply_filters( 'get_space_allowed', $space_allowed );
```
Same question about type casting here.

desrosj commented on src/wp-includes/option.php:
```diff
@@ -1413,64 +1395,37 @@ function get_network_option( $network_id, $option, $default = false ) {
+			 * @since 4.7.0 The `$network_id` parameter was added.
+			 *
+			 */
+			$value = apply_filters( 'default_site_option_' . $option, $default, $option );
```
Looks like this hook is inconsistent in the current state, but dynamic hooks should [use interpolation](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#interpolation-for-naming-dynamic-hooks). Can we update each occurrence to `default_site_option_{$option}`

spacedmonkey replied:
This from the code from in core atm. 

https://github.com/WordPress/wordpress-develop/blob/e94cd298c6e0208a5feb888099601feadf460eda/src/wp-includes/option.php#L1441

spacedmonkey replied:
Fixed.

desrosj commented on src/wp-includes/option.php:
```diff
@@ -1413,64 +1395,37 @@ function get_network_option( $network_id, $option, $default = false ) {
+		if ( is_array( $meta ) && ! empty( $meta ) ) {
+			$value = array_shift( $meta );
+		} else {
+			/**
```
Can we move the hook documentation up to the first occurrence above. I thought that the standard was just the first occurrence, but [looks like it's actually](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#4-1-duplicate-hooks) to "use [svn blame](https://make.wordpress.org/core/handbook/svn/code-history/#using-subversion-annotate) to find the first use of the hook in terms of the earliest revision", which would now be the one within the `if ( ! is_multisite() )` check.

spacedmonkey replied:
Fixed

desrosj commented on src/wp-includes/class-wp-network-query.php:
```diff
@@ -111,27 +111,30 @@ class WP_Network_Query {
 	 *     @type string[]     $path__not_in         Array of paths to exclude affiliated networks for. Default empty.
 	 *     @type string       $search               Search term(s) to retrieve matching networks for. Default empty.
 	 *     @type bool         $update_network_cache Whether to prime the cache for found networks. Default true.
```
These descriptions need to be realigned.

spacedmonkey replied:
Fixed.

desrosj commented on src/wp-includes/ms-network.php:
```diff
@@ -107,35 +108,44 @@ function clean_network_cache( $ids ) {
  * @see update_network_cache()
  * @global wpdb $wpdb WordPress database abstraction object.
  *
  * @param array $network_ids Array of network IDs.
```
These descriptions should also be re-aligned.

spacedmonkey replied:
Fixed

spacedmonkey committed:
Feedback.

spacedmonkey committed:
Update alignment on comments.

spacedmonkey committed:
Merge branch 'trunk' into ix/37181-a

felixarntz commented on src/wp-includes/option.php:
```diff
@@ -1360,19 +1344,17 @@ function update_site_option( $option, $value ) {
  * Retrieves a network's option value based on the option name.
  *
  * @since 4.4.0
+ * @since 6.1.0 Uses get_metadata
```
```suggestion
 * @since 6.1.0 Now uses get_metadata().
```

felixarntz commented on src/wp-includes/option.php:
```diff
@@ -1494,19 +1449,17 @@ function get_network_option( $network_id, $option, $default = false ) {
  * Existing options will not be updated.
  *
  * @since 4.4.0
+ * @since 6.1.0 Uses add_metadata
```
```suggestion
 * @since 6.1.0 Now uses add_metadata().
```

felixarntz commented on src/wp-includes/option.php:
```diff
@@ -1619,8 +1535,10 @@ function add_network_option( $network_id, $option, $value ) {
  * Removes a network option by name.
  *
  * @since 4.4.0
+ * @since 6.1.0 Uses delete_metadata
```
```suggestion
 * @since 6.1.0 Now uses delete_metadata().
```

felixarntz commented on src/wp-includes/option.php:
```diff
@@ -1712,19 +1615,17 @@ function delete_network_option( $network_id, $option ) {
  * Updates the value of a network option that was already added.
  *
  * @since 4.4.0
+ * @since 6.1.0 Uses update_metadata.
```
```suggestion
 * @since 6.1.0 Now uses update_metadata().
```

felixarntz commented on src/wp-includes/option.php:
```diff
@@ -1413,64 +1395,37 @@ function get_network_option( $network_id, $option, $default = false ) {
-				/** This filter is documented in wp-includes/option.php */
-				$value = apply_filters( 'default_site_option_' . $option, $default, $option, $network_id );
-			}
+			/** This action is documented in wp-includes/meta.php */
+			$value = apply_filters( 'default_site_metadata', $value, $network_id, $option, true, 'site' );
```
I'm a bit concerned about calling this here unconditionally. `get_site_option()` allows passing a `$default`, and any developer that uses `default_site_option_{$option}` should be aware of that already, which can be a relevant factor for how to implement the filter callback. For example you could say something like "if the default is `false`, replace it with my custom default here", almost as a shortcut to WP core's `$passed_default` boolean on the regular `default_option_{$option}` filter. (I almost wonder whether we should add that to this filter too.)

Any way, I am diverting a bit: My main concern here is that users of the `default_{$meta_type}_metadata` filters do not consider what the initial parameter value is since there is no way to manually pass a default to metadata functions. So I am wary about developers accidentally overriding this.

This is not necessarily a blocker to committing this, but I would be more comfortable if we only ran this code when no default was manually passed. This way we avoid accidentally modifying a passed default. In addition, it would allow us to simplify this code by actually using the function `get_metadata_default()`.
```suggestion
			// If default was neither passed manually nor modified by the filter, fall back to using metadata default.
			if ( func_num_args() <= 2 && false === $value ) {
				$value = get_metadata_default( 'site', $network_id, $option, true );
			}
```

spacedmonkey replied:
I think if you are going to use the `default_site_metadata` filter, you have to take ownership that. As a developer, it you would have to read the current value and take account of that. 

I do really see where you are coming from here. I think it should be kept how it is and highlighted in the dev note.

spacedmonkey replied:
The sooner this lands the more feedback we can get.

felixarntz reviewed (commented):
@spacedmonkey This looks solid to me, great work!

I mostly left nit-pick comments on missing periods in comments, but I have one comment that deserves more attention: We should be careful about how to handle the defaults, since there are some different implications for network option defaults and metadata defaults - see the relevant comment.

spacedmonkey committed:
Handle numeric values better.

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

spacedmonkey committed:
Remove casting.

spacedmonkey committed:
Merge branch 'ix/37181-a' of github.com:spacedmonkey/wordpress-develop into ix/37181-a

spacedmonkey committed:
Fix unit tests.

spacedmonkey committed:
Move comment.

spacedmonkey commented on src/wp-includes/option.php:
```diff
@@ -1433,42 +1410,22 @@ function get_network_option( $network_id, $option, $default = false ) {
-				$value = apply_filters( 'default_site_option_' . $option, $default, $option, $network_id );
+		$meta = get_metadata_raw( 'site', $network_id, $option );
+		if ( is_array( $meta ) && ! empty( $meta ) ) {
+			$value = array_shift( $meta );
+			if ( is_numeric( $value ) ) {
+				$value = +$value;
```
@peterwilsoncc @felixarntz @desrosj This is a workaround to forces numbric values to be return an int / float. This should help with some edge casts and will mean lots less places where we need to cast to int.

peterwilsoncc replied:
> his is a workaround to forces numbric values to be return an int / float.

Should these be forced to strings instead? For most requests numeric values are returned as strings.

```php
update_site_option( 'pwcc', 21 );
// Same request
var_dump( get_site_option( 'pwcc' ) ); // INT

// Next and subsequent requests
var_dump( get_site_option( 'pwcc' ) ); // String
```

Which is the bug, the first request or the subsequent requests?

spacedmonkey replied:
> subsequent requests

I am not seeing this. 
<img width="382" alt="Screenshot 2022-07-12 at 10 27 43" src="https://user-images.githubusercontent.com/237508/178458379-86742f4d-80eb-4c96-9bd3-52d5d9b866ff.png">

peterwilsoncc replied:
> I am not seeing this.

I'm referring to the current behavior on trunk rather than this branch, ie the change on this branch causes a back-compat break.

![Screen Shot 2022-07-13 at 9 19 03 am](https://user-images.githubusercontent.com/519727/178614526-cd74ea8c-eb57-447c-a32f-74d441044946.png)

I think `get_site_option()` restoring a consistent type is a good thing; I guess the question is whether it is better to break back-compat on the initial or subsequent requests. My feeling is it is best to break the initial request as it's likely the string will be the most frequently returned type.

spacedmonkey replied:
@peterwilsoncc I am not seeing the same results as you. See this screenshot from trunk. 

### Docker
<img width="464" alt="Screenshot 2022-07-19 at 11 05 39" src="https://user-images.githubusercontent.com/237508/179725615-76ed520c-49e8-4674-8f4f-d3f8294cc61a.png">

### Local 
<img width="313" alt="Screenshot 2022-07-19 at 11 19 32" src="https://user-images.githubusercontent.com/237508/179728085-ff4f35c0-4e54-4155-af6d-7dd4bb32181a.png">

I think breaking backwards compat is a really big problem here. Consider the following code. 

```
if ( get_site_option( 'user_count' ) > 1000 ) {
```

or 

```
if ( get_site_option( 'user_count' ) === 1 ) {
```

Which would fail if typing is not correct. It seems like a very common thing to do, so store a number or a count in the site ( network ) option. We can't really break B/C here.

peterwilsoncc replied:
@spacedmonkey The type change to string happens only on subsequent requests (or after exiting and restarting `wp shell`) 

When testing, you need to `exit` prior to `get_site_option()` to see the type change.

I resume this is why core has examples of casting in wrappers for getting options, eg in `get_user_count()`

https://github.com/WordPress/wordpress-develop/blob/2d9073413fa4714d13cc92075fcf4b9d7318777d/src/wp-includes/user.php#L1338

spacedmonkey replied:
@peterwilsoncc OHHHH. I finally get it. Okay, I understand now. Okay, I have added the casting back and tests are passsing. This will need a dev note, but I think we are good.

peterwilsoncc commented:
@spacedmonkey Did you see my note clarifying the backward compatibility break? 

On `trunk` `get_site_option()` always* returns a string but on this branch numeric values are always** returned as an integer.

Notes:
\* Not actually always, see [my comment above](https://github.com/WordPress/wordpress-develop/pull/2478#discussion_r919510207) for the full details
\** always always

spacedmonkey committed:
Revert changes.

spacedmonkey committed:
Cast result.

mukeshpanchal27 commented on src/wp-includes/class-wp-network-query.php:
```diff
@@ -242,8 +245,8 @@ public function get_networks() {
 
-		// Ignore the $fields, $update_network_cache arguments as the queried result will be the same regardless.
-		unset( $_args['fields'], $_args['update_network_cache'] );
+		// Ignore the $fields, $update_network_cache arguments and $update_network_meta_cache as the queried result will be the same regardless.
```
```suggestion
		// Ignore the $fields, $update_network_cache, and $update_network_meta_cache arguments as the queried result will be the same regardless.
```

mukeshpanchal27 reviewed (commented):
Nit-pick.

JJJ committed:
Update src/wp-includes/class-wp-network-query.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

mukeshpanchal27 commented on src/wp-includes/ms-network.php:
```diff
@@ -107,35 +108,44 @@ function clean_network_cache( $ids ) {
  *
- * @param array $networks Array of network row objects.
+ * @param array $networks          Array of network row objects.
+ * @param bool  $update_meta_cache Whether to update sitemeta cache. Default true.
```
Should us `site meta` instead of `sitemeta` as per the existing [document here](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/ms-site.php#L368)
```suggestion
 * @param bool  $update_meta_cache Whether to update site meta cache. Default true.
```

mukeshpanchal27 commented on src/wp-includes/ms-network.php:
```diff
@@ -107,35 +108,44 @@ function clean_network_cache( $ids ) {
  *
- * @param array $network_ids Array of network IDs.
+ * @param array $network_ids       Array of network IDs.
+ * @param bool  $update_meta_cache Whether to update sitemeta cache. Default true.
```
```suggestion
 * @param bool  $update_meta_cache Whether to update site meta cache. Default true.
```

mukeshpanchal27 commented on src/wp-includes/option.php:
```diff
@@ -324,36 +324,20 @@ function wp_load_alloptions( $force_cache = false ) {
  * @since 3.0.0
- *
- * @global wpdb $wpdb WordPress database abstraction object.
+ * @since 6.1.0 Uses update_meta_cache
```
As per the [PHP Documentation Standards](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#since-section-changelogs).

> PHPDoc supports multiple `@since` versions in DocBlocks for this explicit reason. When adding changelog entries to the `@since` block, a version should be cited, and a description should be added in sentence case and form and **end with a period:**

`* @since 6.1.0 Now uses update_meta_cache().`

mukeshpanchal27 replied:
@JJJ Thanks for address other feedback. It seems this change does not added in PR. If i add suggestion for this then it include any third person profile for `@since` tag

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -123,86 +123,256 @@ public function data_network_id_parameter() {
+		$classy->stringy = 'I love slashes\\\\ more';
+		$funky_meta[]    = $classy;
+
+		// Add a network meta item
```
```suggestion
		// Add a network meta item.
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -123,86 +123,256 @@ public function data_network_id_parameter() {
+		// Add a network meta item
+		$this->assertIsInt( add_metadata( 'site', $network_id, $option, $funky_meta, true ) );
+
+		//Check they exists
```
```suggestion
		// Check they exists.
```

mukeshpanchal27 reviewed (commented):
@spacedmonkey Great work!

I mostly left nit-pick comments on docblocks.

JJJ committed:
Update src/wp-includes/ms-network.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update src/wp-includes/ms-network.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Merge branch 'trunk' into ix/37181-a

JJJ committed:
Fix failing test

Corrects the logic inside test_update_network_option_array_with_object(), which I broke when correcting a merge conflict with trunk.

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -137,58 +137,118 @@ public function data_network_id_parameter() {
 
 	/**
-	 * @ticket 43506
+	 * @ticket 37181
```
```suggestion
	 * @ticket 37181
	 *
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -137,58 +137,118 @@ public function data_network_id_parameter() {
-			wp_cache_set( $notoptions_key, $original_cache, 'site-options' );
-		}
+	/**
+	 * @ticket 37181
```
```suggestion
	 * @ticket 37181
	 *
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -137,58 +137,118 @@ public function data_network_id_parameter() {
 
-		$this->assertSame( array(), $cache );
+	/**
+	 * @ticket 37181
```
```suggestion
	 * @ticket 37181
	 *
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -137,58 +137,118 @@ public function data_network_id_parameter() {
 
 	/**
-	 * @ticket 43506
+	 * @ticket 37181
```
```suggestion
	 * @ticket 37181
	 *
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -137,58 +137,118 @@ public function data_network_id_parameter() {
-			wp_cache_delete( $notoptions_key, 'site-options' );
-		}
+	/**
+	 * @ticket 37181
```
```suggestion
	 * @ticket 37181
	 *
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -137,58 +137,118 @@ public function data_network_id_parameter() {
-			wp_cache_set( $notoptions_key, $original_cache, 'site-options' );
-		}
+	/**
+	 * @ticket 37181
```
```suggestion
	 * @ticket 37181
	 *
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -208,24 +270,131 @@ public function test_update_network_option_array_with_object() {
-		// Add the option, it did not exist before this.
-		add_network_option( null, 'array_w_object', $array_w_object );
+	/**
+	 * @dataProvider data_types_options
+	 * @covers ::add_network_option
+	 * @ticket 37181
+	 * @group ms-required
```
```suggestion
	 * @ticket 37181
	 *
	 * @group ms-required
	 *
	 * @covers ::add_network_option
	 *
	 * @dataProvider data_types_options
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -208,24 +270,131 @@ public function test_update_network_option_array_with_object() {
-		// Update the option using the same array with an object for the value.
-		$this->assertFalse( update_network_option( null, 'array_w_object', $array_w_object_2 ) );
+	/**
+	 * @dataProvider data_slashed_options
+	 * @covers ::add_network_option
+	 * @ticket 37181
```
```suggestion
	 * @ticket 37181
	 *
	 * @covers ::add_network_option
	 *
	 * @dataProvider data_slashed_options
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -208,24 +270,131 @@ public function test_update_network_option_array_with_object() {
+	}
+
+	/**
+	 * @dataProvider data_slashed_options
+	 * @covers ::update_network_option
+	 * @ticket 37181
```
```suggestion
	 * @ticket 37181
	 *
	 * @covers ::update_network_option
	 *
	 * @dataProvider data_slashed_options
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -208,24 +270,131 @@ public function test_update_network_option_array_with_object() {
+	}
+
+	/**
+	 * @dataProvider data_slashed_options
+	 * @covers ::delete_network_option()
+	 * @ticket 37181
```
```suggestion
	 * @ticket 37181
	 *
	 * @covers ::delete_network_option()
	 *
	 * @dataProvider data_slashed_options
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -199,6 +259,8 @@ public function test_get_network_option_sets_notoptions_if_option_not_found() {
 	 * @covers ::update_network_option
```
This change will solve the unit test issue.

```suggestion
	 * @group ms-required
         *
	 * @covers ::update_network_option
```

mukeshpanchal27 reviewed (commented):
Per [56070](https://core.trac.wordpress.org/ticket/56070), we should follow a consistent order of annotations in the test suite and https://github.com/WordPress/wordpress-develop/pull/2478#discussion_r959372161 fix the unit test issue.

https://github.com/WordPress/wordpress-develop/pull/2478#discussion_r958377516 is missing from previous review.

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ committed:
Update tests/phpunit/tests/option/networkOption.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

JJJ commented:
tyvm @mukeshpanchal27 💪

(I could've batched these commits, but I wanted to review them manually for my own sake while using the GitHub.com UI to view the diffs.)

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -137,68 +137,138 @@ public function data_network_id_parameter() {
-	 * @covers ::update_network_option
+	 * @group ms-required
+         *
+	 * @Covers ::update_network_option
```
Fix PHPCS issue.

```suggestion
	 * @group ms-required
	 *
	 * @covers ::update_network_option
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -208,24 +278,140 @@ public function test_update_network_option_array_with_object() {
+	 *
+	 * @group ms-required
+	 *
+	 * @Covers ::add_network_option
```
```suggestion
	 * @covers ::add_network_option
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -208,24 +278,140 @@ public function test_update_network_option_array_with_object() {
+	/**
+	 * @ticket 37181
+	 *
+	 * @Covers ::add_network_option
```
```suggestion
	 * @covers ::add_network_option
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -208,24 +278,140 @@ public function test_update_network_option_array_with_object() {
+	/**
+	 * @ticket 37181
+	 *
+	 * @Covers ::update_network_option
```
```suggestion
	 * @covers ::update_network_option
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -208,24 +278,140 @@ public function test_update_network_option_array_with_object() {
+	/**
+	 * @ticket 37181
+	 *
+	 * @Covers ::delete_network_option()
```
```suggestion
	 * @covers ::delete_network_option()
```

mukeshpanchal27 commented on tests/phpunit/tests/option/networkOption.php:
```diff
@@ -208,24 +278,140 @@ public function test_update_network_option_array_with_object() {
-			),
-		);
+		add_metadata( 'site', $network_id, $option, $array_w_object, true );
+		$this->assertSame( $array_w_object, get_network_option( $network_id, $option ) );
```
We should use `assertEquals` instead of `assertSame` similar to `test_get_scheduled_event_singles()` here https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/cron.php#L602-L635 because object key value does not match.

```suggestion
		$this->assertEquals( $array_w_object, get_network_option( $network_id, $option ) );
```

mukeshpanchal27 reviewed (commented):
Thanks @JJJ

Left some reviews that fix the PHPCS and other small capitalized issues, and also address the unit test issue for multisite.

JJJ committed:
Apply suggestions from PHPCS

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

mukeshpanchal27 reviewed (approved):
LGTM.

spacedmonkey committed:
Merge branch 'trunk' into ix/37181-a

spacedmonkey committed:
Merge branch 'trunk' into ix/37181-a

spacedmonkey commented:
https://github.com/WordPress/wordpress-develop/commit/bfbbf4928fc7f075ef4489ed97adbfd9c948d015 Committed
